contracts/SmartLoan.sol

```diff
@@ -0,0 +1,459 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.2;
+
+import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "./IAssetsExchange.sol";
+import "./SupportedAssets.sol";
+import "./Pool.sol";
+import "redstone-flash-storage/lib/contracts/message-based/PriceAwareUpgradeable.sol";
+
+
+/**
+ * @title SmartLoan
+ * A contract that is authorised to borrow funds using delegated credit.
+ * It maintains solvency calculating the current value of assets and borrowings.
+ * In case the value of assets held drops below certain level, part of the funds may be forcibly repaid.
+ * It permits only a limited and safe token transfer.
+ *
+ */
+contract SmartLoan is OwnableUpgradeable, PriceAwareUpgradeable, ReentrancyGuardUpgradeable {
+
+  uint256 public constant PERCENTAGE_PRECISION = 1000;
+
+  uint256 public constant LIQUIDATION_BONUS = 100;
+  uint256 private constant LIQUIDATION_CAP = 200;
+
+  uint256 public MAX_LTV = 5000;
+  uint256 public MIN_SELLOUT_LTV = 4000;
+
+  SupportedAssets supportedAssets;
+  IAssetsExchange public exchange;
+  Pool pool;
+  address private governor;
+
+  function initialize(SupportedAssets _supportedAssets, IAssetsExchange assetsExchange_, Pool pool_, address _governor) external initializer {
+    supportedAssets = _supportedAssets;
+    exchange = assetsExchange_;
+    pool = pool_;
+    __Ownable_init();
+    __PriceAware_init();
+    governor = _governor;
+    MAX_LTV = 5000;
+    MIN_SELLOUT_LTV = 4000;
+  }
+
+
+  /**
+   * Funds a loan with the value attached to the transaction
+  **/
+  function fund() external payable {
+
+    emit Funded(msg.sender, msg.value, block.timestamp);
+  }
+
+
+  function setMaxLTV(uint256 _newMaxLtv) external {
+    require(msg.sender == governor, "Only the governor account can change the maximal LTV");
+    MAX_LTV = _newMaxLtv;
+  }
+
+
+  function setMinSelloutLTV(uint256 _newMinSelloutLtv) external {
+    require(msg.sender == governor, "Only the governor account can change the minimal sellout ltv");
+    MIN_SELLOUT_LTV = _newMinSelloutLtv;
+  }
+
+
+  /**
+   * This function allows selling assets without checking if the loan will remain solvent after this operation.
+   * It is used as part of the sellout() function which sells part/all of assets in order to bring the loan back to solvency.
+   * It is possible that multiple different assets will have to be sold and for that reason we do not use the remainsSolvent modifier.
+  **/
+  function nonSolventAssetSale(bytes32 asset, uint256 _amount, uint256 _minAvaxOut) private {
```

Comment:
The name looks too complex, how about simply sellAsset?

Reply:
I guess you have a point. The thinking behind it was that I wanted to be explicit about which functions do not guarantee that their outcome will be solvent. But I think that just the absence of the `successfulSellout` modifier is enough and this will indeed make the code more readable. Done.